crates/ruff_python_ast/src/source_code/line_index.rs

```diff
@@ -0,0 +1,418 @@
+        let line_range = self.line_range(line_index, contents);
+
+        let column_offset = match self.kind() {
+            IndexKind::Ascii => TextSize::try_from(location.column()).unwrap(),
+            IndexKind::Utf8 => {
+                let line = &contents[line_range];
+
+                // Skip the bom character
+                let bom_len =
+                    usize::from(line_index.to_zero_indexed() == 0 && line.starts_with('\u{feff}'));
+
+                match line.char_indices().nth(location.column() + bom_len) {
+                    Some((offset, _)) => TextSize::try_from(offset).unwrap(),
+                    None => line_range.len(),
+                }
+            }
+        };
+
+        line_range.start() + column_offset
+    }
+
+    /// Return the number of lines in the source code.
+    pub(crate) fn lines_count(&self) -> usize {
+        self.line_starts().len()
+    }
+
+    /// Returns the [byte offset](TextSize) for the `line` with the given index.
+    fn line_start(&self, line: OneIndexed, contents: &str) -> TextSize {
+        let row_index = line.to_zero_indexed();
+        let starts = self.line_starts();
+
+        // If start-of-line position after last line
+        if row_index == starts.len() {
+            contents.text_len()
+        } else {
+            starts[row_index]
+        }
+    }
+
+    /// Returns the [`TextRange`] of the `line` with the given index.
+    /// The start points to the first character's [byte offset](TextSize), the end up to, and including
+    /// the newline character ending the line (if any).
+    fn line_range(&self, line: OneIndexed, contents: &str) -> TextRange {
+        let starts = self.line_starts();
+
+        if starts.len() == line.to_zero_indexed() {
+            TextRange::empty(contents.text_len())
+        } else {
+            TextRange::new(
+                self.line_start(line, contents),
+                self.line_start(line.saturating_add(1), contents),
+            )
+        }
+    }
+
+    /// Returns the [byte offsets](TextSize) for every line
+    pub fn line_starts(&self) -> &[TextSize] {
+        &self.inner.line_starts
+    }
+}
+
+impl Deref for LineIndex {
+    type Target = [TextSize];
+
+    fn deref(&self) -> &Self::Target {
+        self.line_starts()
+    }
+}
+
+impl Debug for LineIndex {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.line_starts()).finish()
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+enum IndexKind {
+    /// Optimized index for an ASCII only document
+    Ascii,
+
+    /// Index for UTF8 documents
+    Utf8,
+}
+
+/// Type-safe wrapper for a value whose logical range starts at `1`, for
+/// instance the line or column numbers in a file
+///
+/// Internally this is represented as a [`NonZeroUsize`], this enables some
+/// memory optimizations
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct OneIndexed(NonZeroUsize);
```

Comment:
I've never used `NonZeroUSize` so I'll just ask: why do we need `OneIndexed` at all? Can we not use that type directly?

Reply:
We could. But I prefer to have semantic wrapper types because they communicate intent and allow us to implement helper methods that are only meaningful in this context (e.g. `to_zero_indexed`) or use method names that are more adequate in this context. 

The use of wrapper types further allows to replace the internal representation if we want (e.g. use a `NonZeroU32`) or to search for all usages of that concept (we may want to use `NonZeroUsize` in other places that are entirely unrelated).